Invalid_argument "index out of bounds"

**Describe the bug**
Scanning Go files, I regularly run into this error. However, the problem is intermittent - which file causes the error changes and sometimes no file causes the error. 

```
Scanning 2753 files with 508 go rules.
Error:  Fatal error at line internal/service/elastictranscoder/pipeline_test.go:1:
 (Invalid_argument "index out of bounds")
====[ BEGIN error trace ]====
Raised by primitive operation at Parse_info.full_charpos_to_pos_large2.(fun) in file "src/pfff/h_program-lang/Parse_info.ml", line 611, characters 24-32
Called from Match_search_rules.regexp_matcher.(fun) in file "src/engine/Match_search_rules.ml", line 452, characters 28-60
Called from Common.fast_map in file "src/pfff/commons/Common.ml", line 428, characters 14-17
Called from Match_search_rules.matches_of_matcher.(fun) in file "src/engine/Match_search_rules.ml", line 325, characters 30-70
Called from Common.fast_map in file "src/pfff/commons/Common.ml", line 431, characters 16-19
Called from Match_search_rules.matches_of_matcher.(fun) in file "src/engine/Match_search_rules.ml", line 323, characters 14-676
Called from Common.with_time in file "src/pfff/commons/Common.ml", line 215, characters 12-16
Called from Match_search_rules.matches_of_matcher in file "src/engine/Match_search_rules.ml", line 322, characters 10-745
Called from Match_search_rules.matches_of_xpatterns in file "src/engine/Match_search_rules.ml", line 589, characters 6-49
Called from Match_search_rules.matches_of_formula in file "src/engine/Match_search_rules.ml", line 1023, characters 4-55
Called from Match_search_rules.nested_formula_has_matches in file "src/engine/Match_search_rules.ml", line 834, characters 4-80
Called from Common.finalize in file "src/pfff/commons/Common.ml", line 619, characters 14-18
Re-raised at Common.finalize in file "src/pfff/commons/Common.ml", line 624, characters 4-11
Called from Stdlib__list.find_all.find in file "list.ml", line 242, characters 17-20
Called from Stdlib__list.fold_left in file "list.ml", line 121, characters 24-34
Called from Match_search_rules.evaluate_formula in file "src/engine/Match_search_rules.ml", line 949, characters 12-96
Called from Match_search_rules.matches_of_formula in file "src/engine/Match_search_rules.ml", line 1039, characters 21-61
Called from Match_search_rules.check_rule in file "src/engine/Match_search_rules.ml", line 1054, characters 4-62
Called from Common.set_timeout in file "src/pfff/commons/Common.ml", line 1429, characters 12-16
Re-raised at Common.set_timeout in file "src/pfff/commons/Common.ml", line 1446, characters 6-13
Called from Match_rules.timeout_function in file "src/engine/Match_rules.ml", line 49, characters 4-73
Called from Match_rules.check.(fun) in file "src/engine/Match_rules.ml", line 109, characters 19-792
Called from Common.partition_either.part_either in file "src/pfff/commons/Common.ml", line 964, characters 15-18
Called from Match_rules.check in file "src/engine/Match_rules.ml", line 89, characters 4-1023
Called from Run_semgrep.semgrep_with_rules.(fun) in file "src/runner/Run_semgrep.ml", line 500, characters 13-265
Called from Run_semgrep.iter_targets_and_get_matches_and_exn_to_errors.(fun) in file "src/runner/Run_semgrep.ml", line 311, characters 21-29
Called from Stdlib__fun.protect in file "fun.ml", line 33, characters 8-15
Re-raised at Stdlib__fun.protect in file "fun.ml", line 38, characters 6-52
Called from Memory_limit.run_with_memory_limit in file "src/system/Memory_limit.ml", line 83, characters 6-62
Called from Run_semgrep.iter_targets_and_get_matches_and_exn_to_errors.(fun) in file "src/runner/Run_semgrep.ml", line 302, characters 17-1023
=====[ END error trace ]=====
```

**To Reproduce**
Using [this config](https://github.com/hashicorp/terraform-provider-aws/blob/main/.semgrep-service-name.yml), scan 2753 Go files. It doesn't _always_ happen.

**Expected behavior**
No error trace, exit `0`.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**What is the priority of the bug to you?**

- [x] P0: blocking your adoption of Semgrep or workflow
- [x] P1: important to fix or quite annoying
- [ ] P2: regular bug that should get fixed

**Environment**
The same thing happens using Docker and an official build. It happens more in a Github Action than in does locally on a MBP but happens in both places.

**Use case**
Continue scanning Go files for consistent code.

I was able to reproduce by running twice `semgrep -c .semgrep-service-name0.yml`

From the repo: https://github.com/hashicorp/terraform-provider-aws


Here is the output of the failure:
```bash
Scanning 3968 files with 107 go rules.
  100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉|3967/3968 tasks
[ERROR] Fatal error at line internal/service/ec2/vpc_endpoint_test.go:1:
 (Invalid_argument "index out of bounds")
====[ BEGIN error trace ]====
Raised by primitive operation at Parse_info.full_charpos_to_pos_large2.(fun) in file "src/pfff/h_program-lang/Parse_info.ml", line 611, characters 24-32
Called from Match_search_rules.regexp_matcher.(fun) in file "src/engine/Match_search_rules.ml", line 470, characters 28-60
Called from Common.fast_map in file "src/pfff/commons/Common.ml", line 428, characters 14-17
Called from Match_search_rules.matches_of_matcher.(fun) in file "src/engine/Match_search_rules.ml", line 325, characters 30-70
Called from Common.fast_map in file "src/pfff/commons/Common.ml", line 430, characters 16-19
Called from Match_search_rules.matches_of_matcher.(fun) in file "src/engine/Match_search_rules.ml", line 323, characters 14-676
Called from Common.with_time in file "src/pfff/commons/Common.ml", line 215, characters 12-16
Called from Match_search_rules.matches_of_matcher in file "src/engine/Match_search_rules.ml", line 322, characters 10-745
Called from Match_search_rules.matches_of_xpatterns in file "src/engine/Match_search_rules.ml", line 607, characters 6-49
Called from Match_search_rules.matches_of_formula in file "src/engine/Match_search_rules.ml", line 1044, characters 4-55
Called from Match_search_rules.nested_formula_has_matches in file "src/engine/Match_search_rules.ml", line 855, characters 4-80
Called from Common.finalize in file "src/pfff/commons/Common.ml", line 619, characters 14-18
Re-raised at Common.finalize in file "src/pfff/commons/Common.ml", line 624, characters 4-11
Called from Stdlib__list.find_all.find in file "list.ml", line 242, characters 17-20
Called from Stdlib__list.fold_left in file "list.ml", line 121, characters 24-34
Called from Match_search_rules.evaluate_formula in file "src/engine/Match_search_rules.ml", line 970, characters 12-96
Called from Match_search_rules.matches_of_formula in file "src/engine/Match_search_rules.ml", line 1060, characters 21-61
Called from Match_search_rules.check_rule in file "src/engine/Match_search_rules.ml", line 1075, characters 4-62
Called from Common.set_timeout in file "src/pfff/commons/Common.ml", line 1434, characters 12-16
Re-raised at Common.set_timeout in file "src/pfff/commons/Common.ml", line 1451, characters 6-13
Called from Match_rules.timeout_function in file "src/engine/Match_rules.ml", line 49, characters 4-73
Called from Match_rules.check.(fun) in file "src/engine/Match_rules.ml", line 109, characters 19-792
Called from Common.partition_either.part_either in file "src/pfff/commons/Common.ml", line 969, characters 15-18
Called from Match_rules.check in file "src/engine/Match_rules.ml", line 89, characters 4-1023
Called from Run_semgrep.semgrep_with_rules.(fun) in file "src/runner/Run_semgrep.ml", line 506, characters 13-265
Called from Run_semgrep.iter_targets_and_get_matches_and_exn_to_errors.(fun) in file "src/runner/Run_semgrep.ml", line 316, characters 21-29
Called from Stdlib__fun.protect in file "fun.ml", line 33, characters 8-15
Re-raised at Stdlib__fun.protect in file "fun.ml", line 38, characters 6-52
Called from Memory_limit.run_with_memory_limit in file "src/system/Memory_limit.ml", line 83, characters 6-62
Called from Run_semgrep.iter_targets_and_get_matches_and_exn_to_errors.(fun) in file "src/runner/Run_semgrep.ml", line 307, characters 17-1023
=====[ END error trace ]=====


Some files were skipped or only partially analyzed.
  Scan was limited to files tracked by git.
  Partially scanned: 1 files only partially analyzed due to a parsing or internal Semgrep error
  Scan skipped: 1613 files matching .semgrepignore patterns
  For a full list of skipped files, run semgrep with the --verbose flag.

Ran 107 rules on 3968 files: 0 findings.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -23,6 +23,7 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
 - Go: single pattern field can now match toplevel fields in a composite
   literal (#5452)
 - PHP: metavariable-pattern: works again when used with language: php (#5443)
+- Fixed a non-deterministic crash when matching a large number of regexes (#5277)
 
 ## [0.97.0](https://github.com/returntocorp/semgrep/releases/tag/v0.97.0) - 2022-06-08
 

--- a/semgrep-core/src/engine/Metavariable_pattern.ml
+++ b/semgrep-core/src/engine/Metavariable_pattern.ml
@@ -89,35 +89,46 @@ let satisfies_metavar_pattern_condition nested_formula_has_matches env r mvar
                   let content = Range.content_at_range mval_file mval_range in
                   Common2.with_tmp_file ~str:content ~ext:"mvar-pattern"
                     (fun file ->
-                      (* We don't want having to re-parse `content', but then we
-                       * need to fix the token locations in `mast`. *)
-                      let mast_start_loc =
-                        mval |> MV.ii_of_mval |> Visitor_AST.range_of_tokens
-                        |> fst |> PI.unsafe_token_location_of_info
-                      in
-                      let fix_loc loc =
-                        {
-                          loc with
-                          PI.charpos = loc.PI.charpos - mast_start_loc.charpos;
-                          line = loc.line - mast_start_loc.line + 1;
-                          column = loc.column - mast_start_loc.column;
-                          file;
-                        }
-                      in
-                      let fixing_visitor =
-                        Map_AST.mk_fix_token_locations fix_loc
-                      in
-                      let mast' = fixing_visitor.Map_AST.vprogram mast in
-                      let xtarget =
-                        {
-                          env.xtarget with
-                          file;
-                          lazy_ast_and_errors = lazy (mast', []);
-                          lazy_content = lazy content;
-                        }
-                      in
-                      nested_formula_has_matches { env with xtarget } formula
-                        (Some r')))
+                      Fun.protect
+                        ~finally:(fun () ->
+                          (* Invalidate the cached line/col lookup table when
+                           * this tmp file is no longer in use.
+                           * https://github.com/returntocorp/semgrep/issues/5277
+                           *
+                           * TODO remove the cache or do more principled cache
+                           * invalidation. *)
+                          Hashtbl.remove Xpattern_matcher.hmemo file)
+                        (fun () ->
+                          (* We don't want having to re-parse `content', but then we
+                           * need to fix the token locations in `mast`. *)
+                          let mast_start_loc =
+                            mval |> MV.ii_of_mval |> Visitor_AST.range_of_tokens
+                            |> fst |> PI.unsafe_token_location_of_info
+                          in
+                          let fix_loc loc =
+                            {
+                              loc with
+                              PI.charpos =
+                                loc.PI.charpos - mast_start_loc.charpos;
+                              line = loc.line - mast_start_loc.line + 1;
+                              column = loc.column - mast_start_loc.column;
+                              file;
+                            }
+                          in
+                          let fixing_visitor =
+                            Map_AST.mk_fix_token_locations fix_loc
+                          in
+                          let mast' = fixing_visitor.Map_AST.vprogram mast in
+                          let xtarget =
+                            {
+                              env.xtarget with
+                              file;
+                              lazy_ast_and_errors = lazy (mast', []);
+                              lazy_content = lazy content;
+                            }
+                          in
+                          nested_formula_has_matches { env with xtarget }
+                            formula (Some r'))))
           | Some xlang, MV.Text (content, _tok)
           | Some xlang, MV.Xmls [ XmlText (content, _tok) ]
           | Some xlang, MV.E { e = G.L (G.String (content, _tok)); _ } ->
@@ -127,37 +138,48 @@ let satisfies_metavar_pattern_condition nested_formula_has_matches env r mvar
               (* We re-parse the matched text as `xlang`. *)
               Common2.with_tmp_file ~str:content ~ext:"mvar-pattern"
                 (fun file ->
-                  let lazy_ast_and_errors =
-                    lazy
-                      (match xlang with
-                      | L (lang, _) ->
-                          let { Parse_target.ast; skipped_tokens; _ } =
-                            Parse_target.parse_and_resolve_name lang file
-                          in
-                          (* TODO: If we wanted to report the parse errors
-                           * then we should fix the parse info with
-                           * Parse_info.adjust_info_wrt_base! *)
-                          if skipped_tokens <> [] then
-                            pr2
-                              (spf
-                                 "rule %s: metavariable-pattern: failed to \
-                                  fully parse the content of %s"
-                                 (fst env.rule.Rule.id) mvar);
-                          (ast, skipped_tokens)
-                      | LRegex
-                      | LGeneric ->
-                          failwith "requesting generic AST for LRegex|LGeneric")
-                  in
-                  let xtarget =
-                    {
-                      Xtarget.file;
-                      xlang;
-                      lazy_ast_and_errors;
-                      lazy_content = lazy content;
-                    }
-                  in
-                  nested_formula_has_matches { env with xtarget } formula
-                    (Some r'))
+                  Fun.protect
+                    ~finally:(fun () ->
+                      (* Invalidate the cached line/col lookup table when this
+                       * tmp file is no longer in use.
+                       * https://github.com/returntocorp/semgrep/issues/5277
+                       *
+                       * TODO remove the cache or do more principled cache
+                       * invalidation. *)
+                      Hashtbl.remove Xpattern_matcher.hmemo file)
+                    (fun () ->
+                      let lazy_ast_and_errors =
+                        lazy
+                          (match xlang with
+                          | L (lang, _) ->
+                              let { Parse_target.ast; skipped_tokens; _ } =
+                                Parse_target.parse_and_resolve_name lang file
+                              in
+                              (* TODO: If we wanted to report the parse errors
+                               * then we should fix the parse info with
+                               * Parse_info.adjust_info_wrt_base! *)
+                              if skipped_tokens <> [] then
+                                pr2
+                                  (spf
+                                     "rule %s: metavariable-pattern: failed to \
+                                      fully parse the content of %s"
+                                     (fst env.rule.Rule.id) mvar);
+                              (ast, skipped_tokens)
+                          | LRegex
+                          | LGeneric ->
+                              failwith
+                                "requesting generic AST for LRegex|LGeneric")
+                      in
+                      let xtarget =
+                        {
+                          Xtarget.file;
+                          xlang;
+                          lazy_ast_and_errors;
+                          lazy_content = lazy content;
+                        }
+                      in
+                      nested_formula_has_matches { env with xtarget } formula
+                        (Some r')))
           | Some _lang, mval ->
               (* This is not necessarily an error in the rule, e.g. you may be
                * matching `$STRING + ...` and then add a metavariable-pattern on

